_posts/2017-01-23-cpp-strings.md

```diff
@@ -0,0 +1,155 @@
+---
+title: Modeling strings with libpmemobj C++ bindings
+author: RobDickinson
+layout: post
+---
+
+C++ developers using libpmemobj have more than one option for modeling strings,
+depending on the size of the strings and whether they are fixed or varying
+in length. In this post we'll review the representations that work,
+known variations to avoid, and finally present a persistent string
+class that implements these best practices.
+
+#### Avoid wrapping fixed-size arrays
+
+You might expect (like I did at first!) that `p<char[size]>` is a proper way
+to simply model a fixed-size string, but actually this is not correct.
+The `p<>` template does not implement the subscript `operator[]` and so
+`p<char[size]>` won't even compile.
+
+#### Avoid wrapping std::string
+
+Never use `p<std::string>` or `persistent_ptr<std::string>`. This code will
+compile, but these implementations will not be power-fail safe (at best)
+and might be unstable (at worst). This advice applies to using `p<>`
+or `persistent_ptr<>` wrappers around any complex types that perform
+their own memory management. Don't make this obvious mistake!
+
+#### Avoid arrays of persistent pointers to single chars
+
+Don't model a variable-length string using `persistent_ptr<char> X[]`. This
```

Comment:
Hi @tomaszkapela,

I took a little time to really think about what we really want to emphasize here.  The points that you raise are good ones, but I'm struggling to convey all of this in a way that makes this clear.

What I settled on for this latest draft is that the main objection to the 'persistent pointers to single char' is its horrible efficiency. I removed any statements about any of these variations "working" or being "safe" and am just focusing on the efficiency angle as being horrible enough to wave off doing this.

So I know I'm leaving some out details you just suggested for this latest draft, but my thought was one justification is enough for the anti-patterns and we don't need to list every fault.  For the "avoid wrapping fixed-size arrays" case, it won't compile.  For the "avoid wrapping std::string" case, it's unsafe.  For the "avoid arrays of persistent pointers to single char" case, it's ridiculously inefficient.

Anyway, I'm curious to see what you think of this latest version.  And thanks for all the time invested in this so far!